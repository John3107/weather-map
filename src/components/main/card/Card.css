.card {
    margin: 20px 0;
}

.card:first-of-type {
    margin-top: 160px;
}

.paper {
    display: flex;
    flex-direction: column;
    padding: 10px;
}

.main {
    display: inherit;
    justify-content: space-between;
    align-items: center;
}

.main img {
    width: 50px;
    height: 50px;
}

.left-side {
    display: inherit;
    justify-content: space-between;
    align-items: center;
    margin-right: 15px;
}

.left-side div {
    display: inherit;
    flex-direction: column;
}

.right-side {
    display: inherit;
    flex-direction: column;
    align-items: flex-end;
}

.right-side span {
    text-align: right;
}

.actions {
    display: inherit;
    justify-content: flex-end;
}

@media (max-width: 660px) {
    .main {
        align-items: flex-start;
        flex-direction: column;
    }

    .right-side {
        align-items: flex-start;
        margin-top: 20px;
    }

    .right-side span{
        margin-top: 10px;
        text-align: left;
    }

    .actions {
        justify-content: flex-start;
        margin-top: 20px;
    }
}