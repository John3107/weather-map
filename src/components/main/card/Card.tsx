import React, {useEffect, useState} from 'react';
import './Card.css';
import {Paper} from "@material-ui/core";
import {CityWeatherType} from "../../../types/types";
import {Refresh} from '@material-ui/icons';
import {Delete} from '@material-ui/icons';
import {useAppDispatch, useAppSelector} from "../../../hooks/hooks";
import {
    deleteCardTC,
    refreshCityTC,
    selectedCityTC,
    setDeletedCardAC
} from "../../../bll/app-reducer";
import {Navigate, useNavigate} from "react-router-dom";
import CityWeatherInfo from "../../cityWeatherInfo/CityWeatherInfo";

const Card = (props: CityWeatherType) => {
    const data = useAppSelector()
    const dispatch = useAppDispatch()
    const navigate = useNavigate()
    const {name, country, temp, description, icon, id, lon, lat} = props
    const [date, setDate] = useState('')

    useEffect(() => {
        setDate(String(new Date()).split('GMT')[0])
    }, [])

    const deleteCardHandler = () => {
        dispatch(deleteCardTC(id))
        const removeCard = data.cities.filter(el => el.id !== id)
        dispatch(setDeletedCardAC(removeCard))
    }

    const refreshDataHandler = () => {
        setDate(String(new Date()).split('GMT')[0])
        dispatch(refreshCityTC({name, country, lon, lat}))
    }

    const toInfo = () => {
        dispatch(selectedCityTC(props))
        navigate(`/city/${name}`)
        return <Navigate to={`/city/${name}`}/> && <CityWeatherInfo/>
    }

    return (
        <div className="card" data-testid="card-test">
            <Paper elevation={3} className="paper">
                <div className="main" onClick={toInfo}>
                    <div className="left-side">
                        <img src={`http://openweathermap.org/img/wn/${icon}@2x.png`} alt=""/>
                        <div>
                            <span style={{fontSize: '26px'}}>{temp}CÂ°</span>
                            <span>{description}</span>
                        </div>
                    </div>
                    <div className="right-side">
                        <div>
                            <span>{name}, </span>
                            <span>{country}</span>
                        </div>
                        <span>{date}</span>
                    </div>
                </div>
                <div className="actions">
                    <Refresh style={{cursor: 'pointer'}} onClick={refreshDataHandler}/>
                    <Delete style={{cursor: 'pointer'}} onClick={deleteCardHandler}/>
                </div>
            </Paper>
        </div>
    );
}

export default Card;