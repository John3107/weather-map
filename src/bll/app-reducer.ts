import {CityType, CityWeatherType} from "../types/types";
import {AppThunk} from "./store";
import {weatherAPI} from "../api/api";

const initialState: InitialStateType = {
    cities: [],
    citiesContext: []
}

export const appReducer = (state: InitialStateType = initialState, action: ActionsType): InitialStateType => {
    switch (action.type) {
        case 'CITIES':
            return {...state, cities: action.citiesList}
        case 'CITIES-CONTEXT':
            return {...state, citiesContext: action.citiesContextArray}
        default:
            return state
    }
}

export const initialCitiesDataTC = (): AppThunk => (dispatch) => {
    const getCities = localStorage.getItem('cities-list')
    if (getCities) {
        dispatch(setCitiesAC(JSON.parse(getCities)))
    }
}

export const addCityTC = (newCity: CityType): AppThunk => (dispatch) => {
    dispatch(setCitiesContextAC([]))
    const citiesList = localStorage.getItem('cities-list')
    const citiesListParsed = citiesList && JSON.parse(citiesList)
    weatherAPI.getCityWeather(newCity.lat, newCity.lon)
        .then((res) => {
            const cityWeather = {
                id: String(res.data.lon) + String(res.data.lat),
                city: newCity.name,
                country: newCity.country,
                temp: Math.round(res.data.current.temp / 10.59),
                description: res.data.current.weather[0].description,
                icon: res.data.current.weather[0].icon,
                lat: res.data.lat,
                lon: res.data.lon
            }
            if (citiesListParsed[0]) {
                const searchCity = citiesListParsed.find((el: CityWeatherType) => el.id === cityWeather.id)
                let addCity
                if (searchCity) {
                    addCity = citiesListParsed.map((el: CityWeatherType) => el.id === cityWeather.id
                        ? ({...cityWeather})
                        : el)
                } else {
                    addCity = [...citiesListParsed, cityWeather]
                }
                localStorage.setItem('cities-list', JSON.stringify(addCity))
                dispatch(setCitiesAC(addCity))
            } else {
                localStorage.setItem('cities-list', JSON.stringify([cityWeather]))
                dispatch(setCitiesAC([cityWeather]))
            }
        })
        .catch((err) => {
            console.log(err)
        })
}

export const citiesContextTC = (title: string): AppThunk => (dispatch) => {
    if (!title) return dispatch(setCitiesContextAC([]))
    weatherAPI.getCitiesWeather(title)
        .then((res) => {
            dispatch(setCitiesContextAC(res.data))
        })
        .catch((err) => {
            console.log(err)
        })
}

export const deleteCardTC = (cardId: string): AppThunk => (dispatch) => {
    const citiesList = localStorage.getItem('cities-list')
    const citiesListParsed = citiesList && JSON.parse(citiesList)
    const citiesListFiltered = citiesListParsed.filter((el: CityWeatherType) => el.id !== cardId)
    localStorage.setItem('cities-list', JSON.stringify(citiesListFiltered))
    dispatch(setCitiesAC(citiesListFiltered))
}

export const setCitiesAC = (citiesList: CityWeatherType[]) => ({type: 'CITIES', citiesList} as const)
export const setCitiesContextAC = (citiesContextArray: CityType[]) =>
    ({type: 'CITIES-CONTEXT', citiesContextArray} as const)

export type CitiesActionType = ReturnType<typeof setCitiesAC>
export type CitiesContextActionType = ReturnType<typeof setCitiesContextAC>

export type InitialStateType = {
    cities: CityWeatherType[],
    citiesContext: CityType[]
}

type ActionsType = CitiesActionType | CitiesContextActionType