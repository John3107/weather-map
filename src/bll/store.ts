import {AnyAction, applyMiddleware, combineReducers, legacy_createStore as createStore} from 'redux'
import thunkMiddleware, {ThunkAction, ThunkDispatch} from 'redux-thunk'
import {appReducer} from "./app-reducer";

export const rootReducer = combineReducers({
    app: appReducer
})
export const store = createStore(rootReducer, applyMiddleware(thunkMiddleware));

export type AppRootStateType = ReturnType<typeof rootReducer>

export type AppDispatch = ThunkDispatch<AppRootStateType, unknown, AnyAction>

export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppRootStateType, unknown, AnyAction>