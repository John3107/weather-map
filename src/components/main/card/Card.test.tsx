import { Provider } from "react-redux";
// @ts-ignore
import renderer from "react-test-renderer";
import Card from './Card';
import {store} from "../../../bll/store";

it('renders correctly snapshot', () => {
    const tree = renderer
        .create(<Provider store={store}><Card prop={store.getState().app.selectedCity}/></Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});