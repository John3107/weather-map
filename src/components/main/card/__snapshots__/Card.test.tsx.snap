// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="card"
  data-testid="card-test"
>
  <div
    className="MuiPaper-root paper MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      className="main"
    >
      <div
        className="left-side"
      >
        <img
          alt=""
          src="http://openweathermap.org/img/wn/@2x.png"
        />
        <div>
          <span
            style={
              Object {
                "fontSize": "26px",
              }
            }
          >
            0
            C°
          </span>
          <span />
        </div>
      </div>
      <div
        className="right-side"
      >
        <div>
          <span>
            , 
          </span>
          <span />
        </div>
        <span />
      </div>
    </div>
    <div
      className="actions"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders correctly snapshot 1`] = `
<div
  className="card"
  data-testid="card-test"
>
  <div
    className="MuiPaper-root paper MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      className="main"
    >
      <div
        className="left-side"
      >
        <img
          alt=""
          src="http://openweathermap.org/img/wn/@2x.png"
        />
        <div>
          <span
            style={
              Object {
                "fontSize": "26px",
              }
            }
          >
            0
            C°
          </span>
          <span />
        </div>
      </div>
      <div
        className="right-side"
      >
        <div>
          <span>
            , 
          </span>
          <span />
        </div>
        <span />
      </div>
    </div>
    <div
      className="actions"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
</div>
`;
